ldaptive/src/main/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDao.java

```diff
@@ -496,6 +507,125 @@ public void removeFromGroup(User user, Group group) throws DaoException {
         modifyGroupMembership(user, group, AttributeModificationType.REMOVE);
     }
 
+    /**
+     * Returns all available groups
+     *
+     * Note that this method is not particular to a user, but is temporarily made available here until a
+     * more suitable framework becomes available for providing group dao.
+     *
+     * @param baseSearchDn
+     *  null value indicates to return all groups
+     *
+     * @return list of all available groups under base search dn
+     */
+    @Override
+    public List<Group> getAllGroups(String baseSearchDn) throws ExceededMaximumAllowedResultsException {
+        List<Group> groups = Lists.newArrayList();
+
+        try {
+            enqueueAllByFilter(groups, null, baseSearchDn, SEARCH_NO_LIMIT, false);
+        } catch (final ExceededMaximumAllowedResultsException e) {
+            // propagate ExceededMaximumAllowedResultsException exceptions
+            throw e;
+        } catch (final DaoException suppressed) {
+            // suppress any other DaoExceptions
+            return Collections.emptyList();
+        }
+
+        return groups;
+    }
+
+    /**
+     * @param groups                    collection to populate with loaded groups
+     * @param filter                    the LDAP search filter to use when searching
+     * @param baseSearchDn        groups base search dn, null defaults to all groups (this.groupsBaseSearchDn)
+     * @param limit                     the maximum number of results to return, a limit of 0 indicates that all results
+     *                                  should be returned
+     * @param restrictMaxAllowedResults a flag indicating if maxSearchResults should be observed
+     * @return number of users loaded
+     * @throws DaoException
+     */
+    private int enqueueAllByFilter(@Nonnull final Collection<Group> groups, @Nullable BaseFilter filter,
+                                   String baseSearchDn, final int limit, final boolean restrictMaxAllowedResults)
+            throws DaoException {
+
+        filter = filter != null ? filter.and(FILTER_GROUP) : FILTER_GROUP;
+
+        // perform search
+        Connection conn = null;
+        try {
+            conn = this.connectionFactory.getConnection();
+            conn.open();
+            SearchOperation search = new SearchOperation(conn);
+
+            boolean useBaseSearchDn = baseSearchDn != null &&
+                    baseSearchDn.toLowerCase().endsWith(this.groupsBaseSearchDn.toLowerCase());
```

Comment:
should we throw an `IllegalArgumentException` if `!baseSearchDn.toLowerCase().endsWith(this.groupsBaseSearchDn.toLowerCase())`?

that would prevent groups in the wrong baseDn from being returned due to an accidental misconfiguration